### PR TITLE
Created and implemented an automatically passed argument system. 

### PR DESCRIPTION
**Pull request by ALEXWARELLC**
This pull request contains some new changes and additions.

## Additions
- LocalAdmin now starts by merging command-line arguments and arguments contained in `laargs.txt`. This can help with providing a quick & easy way to launch LocalAdmin without having to type arguments each time.

## Changes
- `LocalAdmin V2.csproj` now includes `laargs.txt`.

## Extra Information
- Any line in `laargs.txt` that starts with '#' will be ignored.
- If `laargs.txt` doesn't exist, only the command-line arguments will be passed.
- `laargs.txt` doesn't require the first argument to be the port number.
- It will also support any new arguments as it only passes the arguments located in `laargs.txt` and in the command-line.

That's all.
ALEXWARELLC

### CODE REVIEW

File: Core/StartupArgManager.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            List<string> StartupArgs = new List<string>();
+            foreach (string arg in CMDArgs)
+            {
+                if (!StartupArgs.Contains(arg))
+                    StartupArgs.Add(arg);
+            }
+
+            foreach (string farg in File.ReadAllLines(Path.GetFullPath(StartupArgsPath)))
+            {
+                if (!farg.StartsWith("#") && !farg.StartsWith(" "))
```
Comment (zabszk) [inline Core/StartupArgManager.cs:30]: String comparison should be set to Ordinal for both checks.

File: Core/StartupArgManager.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                {
+                    if (!StartupArgs.Contains(farg))
```
Comment (zabszk) [inline Core/StartupArgManager.cs:32]: This condition should be removed. If someone puts there `--arg1 a --arg2 a` it will become just `--arg1 a --arg2`

File: Core/StartupArgManager.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        public static string[] MergeStartupArgs(string[] CMDArgs)
+        {
+            List<string> StartupArgs = new List<string>();
+            foreach (string arg in CMDArgs)
+            {
+                if (!StartupArgs.Contains(arg))
+                    StartupArgs.Add(arg);
+            }
+
+            foreach (string farg in File.ReadAllLines(Path.GetFullPath(StartupArgsPath)))
```
Comment (zabszk) [inline Core/StartupArgManager.cs:28]: This will throw an exception if the file is not found and prevent entire LA from starting.

Instead of bundling the file together with LA it should be created on startup if it doesn't exist.

**Comment by ALEXWARELLC** [comment]
As long as it works, that's what matters.

https://discord.com/channels/656673194693885975/656673194693885981/1156293300475678872

File: Core/StartupArgManager.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        /// Merges Command-line arguments and arguments in <paramref name="StartupArgsPath"/>
+        /// </summary>
+        /// <param name="CMDArgs">Runtime Command-Line Arguments</param>
+        /// <returns>Merged Arguments</returns>
+        public static string[] MergeStartupArgs(string[] CMDArgs)
+        {
+            List<string> StartupArgs = new List<string>();
+            foreach (string arg in CMDArgs)
+            {
+                if (!StartupArgs.Contains(arg))
```
Comment (zabszk) [inline Core/StartupArgManager.cs:24]: This condition should be removed. If someone puts there `--arg1 a --arg2 a` it will become just `--arg1 a --arg2`.

File: Core/StartupArgManager.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace LocalAdmin.V2.Core
+{
+    public sealed class StartupArgManager
```
Comment (zabszk) [inline Core/StartupArgManager.cs:7]: Class should be static

File: Core/StartupArgManager.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        /// </summary>
+        /// <param name="CMDArgs">Runtime Command-Line Arguments</param>
+        /// <returns>Merged Arguments</returns>
+        public static string[] MergeStartupArgs(string[] CMDArgs)
+        {
+            List<string> StartupArgs = new List<string>();
+            foreach (string arg in CMDArgs)
+            {
+                if (!StartupArgs.Contains(arg))
+                    StartupArgs.Add(arg);
```
Comment (zabszk) [inline Core/StartupArgManager.cs:25]: AddRange can be used instead of this entire foreach loop.

File: Core/StartupArgManager.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        /// Path to startup arguments file.
+        /// </summary>
+        private const string StartupArgsPath = "laargs.txt";
+
+        /// <summary>
+        /// Merges Command-line arguments and arguments in <paramref name="StartupArgsPath"/>
+        /// </summary>
+        /// <param name="CMDArgs">Runtime Command-Line Arguments</param>
+        /// <returns>Merged Arguments</returns>
+        public static string[] MergeStartupArgs(string[] CMDArgs)
```
Comment (zabszk) [inline Core/StartupArgManager.cs:19]: If an exception happens here (and it can happen) we should just return the original array and print the exception.

And this can take `IEnumerable<string>` instead of `string[]` as an arg.

**Review by zabszk — CHANGES_REQUESTED** [review]
Overall good idea.

File: Core/StartupArgManager.cs (hunk 1 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:30]: I can't be bothered trying to figure out how to implement that properly so i'm just ignoring lines that start with a '#'.

File: Core/StartupArgManager.cs (hunk 2 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:32]: Removed

File: Core/StartupArgManager.cs (hunk 3 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:28]: Fixed but it's a bit janky.

File: Core/StartupArgManager.cs (hunk 4 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:24]: Done.

File: Core/StartupArgManager.cs (hunk 5 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:7]: Done.

File: Core/StartupArgManager.cs (hunk 6 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:25]: Done, and changed for CMDArgs.

File: Core/StartupArgManager.cs (hunk 7 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:19]: Done, it's not fool proof, but as long as it works, that's what matters.

**Comment by ALEXWARELLC** [comment]
Imma go sleepy sleep.

File: Core/StartupArgManager.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                foreach (string farg in File.ReadAllLines(Path.GetFullPath(StartupArgsPath)))
+                {
+                    if (farg.StartsWith("#"))
+                        StartupArgs.Add(farg);
+                }
+                return StartupArgs.ToArray();
+            }
+            catch (Exception ex)
+            {
+                ConsoleUtil.WriteLine($"An error occured while trying to merge arguments :: {ex}", ConsoleColor.Red);
```
Comment (zabszk) [inline Core/StartupArgManager.cs:40]:
```suggestion
                ConsoleUtil.WriteLine($"An error occured while trying to merge arguments: {ex}", ConsoleColor.Red);
```

File: Core/StartupArgManager.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        /// <returns>Merged Arguments</returns>
+        public static string[] MergeStartupArgs(IEnumerable<string> CMDArgs)
+        {
+            try
+            {
+                List<string> StartupArgs = new List<string>();
+                StartupArgs.AddRange(CMDArgs);
+
+                if (!File.Exists(Path.GetFullPath(StartupArgsPath)))
+                    File.WriteAllText(Path.GetFullPath(StartupArgsPath), string.Empty);
```
Comment (zabszk) [inline Core/StartupArgManager.cs:29]: No need to run the rest of the code after just creating the file. It should return CMDArgs in that case.

File: Core/StartupArgManager.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+                foreach (string farg in File.ReadAllLines(Path.GetFullPath(StartupArgsPath)))
+                {
+                    if (farg.StartsWith("#"))
```
Comment (zabszk) [inline Core/StartupArgManager.cs:33]: Ordinal comparison mode should be used.

File: Core/StartupArgManager.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+                        StartupArgs.Add(farg);
+                }
+                return StartupArgs.ToArray();
```
Comment (zabszk) [inline Core/StartupArgManager.cs:36]: New line should be added before the return.

File: Properties/launchSettings.json [hunk 12]
```diff
@@ -1,7 +1,8 @@
 {
   "profiles": {
     "LocalAdmin V2 Net Core": {
-      "commandName": "Project"
+      "commandName": "Project",
+      "commandLineArgs": "-s"
```
Comment (zabszk) [inline Properties/launchSettings.json:5]: What's the purpose of this change?

**Review by zabszk — CHANGES_REQUESTED** [review]

File: Core/StartupArgManager.cs (hunk 10 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:33]: Using `StringComparison.Ordinal` now.

File: Core/StartupArgManager.cs (hunk 9 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:29]: Completed.

File: Core/StartupArgManager.cs (hunk 11 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs:36]: Added new line.

File: Properties/launchSettings.json (hunk 12 above)
Comment (ALEXWARELLC) [reply Properties/launchSettings.json:5]: Removed.

**Comment by ALEXWARELLC** [comment]
Wish

**Comment by ALEXWARELLC** [comment]
Review?

File: Core/StartupArgManager.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+                List<string> StartupArgs = new List<string>();
+                StartupArgs.AddRange(CMDArgs);
+
+                if (!File.Exists(Path.GetFullPath(StartupArgsPath)))
+                {
+                    File.WriteAllText(Path.GetFullPath(StartupArgsPath), string.Empty);
+                    return CMDArgs.ToArray();
+                }
+
+                foreach (string farg in File.ReadAllLines(Path.GetFullPath(StartupArgsPath)))
```
Comment (zabszk) [inline Core/StartupArgManager.cs:34]: I don't think `Path.GetFullPath` is needed (on lines 28, 30 and 34). It should work fine with only providing the file name.

**Review by zabszk — CHANGES_REQUESTED** [review]

File: Core/StartupArgManager.cs (hunk 13 above)
Comment (ALEXWARELLC) [reply Core/StartupArgManager.cs]: This has been updated and has been resolved.

**Comment by ALEXWARELLC** [comment]
Update: Tested and working.

**Review by zabszk — APPROVED** [review]
